File: BDDExamplesSwiftTests/PhotoUploaderVCSpec.m
//
//  PhotoUploaderVCSpec.m
//  BDDExamplesSwift
//
//  Created by huoshuguang on 2017/5/1.
//  Copyright © 2017年 huoshuguang. All rights reserved.
//

@import OCMock;
@import Quick;
@import Nimble;
#import "UIBarButtonItem+Specs.h"
#import "BDDExamplesSwiftTests-Swift.h"
/**/
//@interface PhotoUploadViewController (Specs)
//- (void)didTapUploadButton:(UIBarButtonItem *)uploadButton;
//@end

QuickSpecBegin(PhotoUploaderVCSpecd)

describe(@"PhotoUploadViewController", ^{
    __block PhotoUploadViewController *photoUploadViewController;
    __block id mockPhotoUploader;
    
    beforeEach(^{
        mockPhotoUploader = OCMClassMock([PhotoUploader class]);
        photoUploadViewController = [[PhotoUploadViewController alloc] initWithUploader:mockPhotoUploader];
        [photoUploadViewController viewDidLoad];
    });
    
    afterEach(^{
        photoUploadViewController = nil;
    });
    
    // Non-behavior style
    
    describe(@"right bar button item", ^{
        
        __block UIBarButtonItem *barButtonItem;
        
        beforeEach(^{
            barButtonItem = [[photoUploadViewController navigationItem] rightBarButtonItem];
        });
        
        it(@"should have a title", ^{
            expect(barButtonItem.title).to(equal(@"Upload"));
        });
        
        it(@"should have a target", ^{
            expect(barButtonItem.target).to(equal(photoUploadViewController));
        });
        
        it(@"should have an action", ^{
            //expect(barButtonItem.action).to(equal(@selector(didTapUploadButton:)));
        });
    });
    
    describe(@"tapping right bar button item", ^{
        beforeEach(^{
            [photoUploadViewController didTapUploadButtonWithUploadButton:[UIBarButtonItem new]];
        });
        
        it(@"should tell the mock photo uploader to upload the photo", ^{
//            [verify(mockPhotoUploader) uploadPhoto:instanceOf([UIImage class])
//                                        completion:anything()];
            
            OCMVerify([(PhotoUploader *)mockPhotoUploader upLoadWithImage:[UIImage new] to:[[NSURL alloc] initWithString:@""] handler:^(NSString *result,NSString *error) {
                NSLog(@"======");
            }]);
        });
    });
    
    // BDD Style
    
    describe(@"right bar button item", ^{
        
        __block UIBarButtonItem *barButtonItem;
        
        beforeEach(^{
            barButtonItem = [[photoUploadViewController navigationItem] rightBarButtonItem];
        });
        
        it(@"should have a title", ^{
            expect(barButtonItem.title).to(equal(@"Upload"));
        });
        
        describe(@"when it is tapped", ^{
            beforeEach(^{
                [barButtonItem specsSimulateTap];
            });
            
            it(@"should tell the mock photo uploader to upload the photo", ^{
//                [verify(mockPhotoUploader) uploadPhoto:instanceOf([UIImage class])
//                                            completion:anything()];
                PhotoUploader *uploader = (PhotoUploader *)mockPhotoUploader;
                OCMVerify([uploader upLoadWithImage:[UIImage new] to:[[NSURL alloc] initWithString:@""] handler:^(NSString *result,NSString *error) {
                    NSLog(@"-------0000000000");
                    
                }]);
            });
        });
    });
});


QuickSpecEnd

File: BDDExamplesSwiftTests/PhotoUploaderVCSpec.swift
//
//  PhotoUploaderVCSpec.swift
//  BDDExamplesSwift
//
//  Created by huoshuguang on 2017/5/1.
//  Copyright © 2017年 huoshuguang. All rights reserved.
//

import Quick
import Nimble
@testable import BDDExamplesSwift



class PhotoUploaderVCSpec: QuickSpec
{
    override func spec()
    {
        describe("PhotoUploadViewController") { 
            var photoUploadViewController:PhotoUploadViewController!
            //var mockPhotoUploader:PhotoUploader!
            beforeEach {
                //
//                mockPhotoUploader = OCMStrictClassMock(for: PhotoUploader.self) as! PhotoUploader
                let loader = PhotoUploader.init()
                photoUploadViewController = PhotoUploadViewController.init(uploader: loader)
                photoUploadViewController.viewDidLoad()
            }
            afterEach {
                photoUploadViewController = nil
            }
            
            describe("right bar button item", { 
                var barButtonItem:UIBarButtonItem!
                
                beforeEach {
                    //
                    barButtonItem = photoUploadViewController.navigationItem.rightBarButtonItem
                }
                
                it("should have a title", closure: { 
                    expect(barButtonItem.title).to(equal("Upload"))
                })
                
                it("should have a target", closure: { 
                    //
                    expect(barButtonItem.target).to(be(photoUploadViewController))
                })
                
                it("should have an action", closure: {
                    //be()
                    let action:Selector = #selector(PhotoUploadViewController.didTapUploadButton(uploadButton:))
                    expect(barButtonItem.action).to(be(action))
                })
            })
            
            describe("tapping right bar button item", { 
                beforeEach {
//                    photoUploadViewController.didTapUploadButton(uploadButton: nil)
                }
                
                it("should tell the mock photo uploader to upload the photo", closure: {
                   
                    //
//                    OCMVerify(mockPhotoUploader.upLoad(image: UIImage.init(), to: NSURL.init(string: "")!, handler: { (string, tring) in
//                        print("----")
//                    }))
                })
            })
        }
    }
}
